File: services/python/repositories/user_repository.py
from typing import Optional
from .base_repository import BaseRepository
from ..models.user import User

class UserRepository(BaseRepository[User]):
    """User repository"""
    
    def _get_model_class(self):
        return User

    def _get_table_name(self):
        return 'users'

    def _build_model(self, row: dict) -> User:
        return User(
            id=row['id'],
            email=row['email'],
            password_hash=row['password_hash'],
            created_at=row['created_at'],
            updated_at=row['updated_at']
        )

    def find_by_email(self, email: str) -> Optional[User]:
        """Find user by email"""
        query = "SELECT * FROM users WHERE email = %s"
        cursor = self.db.cursor()
        cursor.execute(query, (email,))
        row = cursor.fetchone()
        cursor.close()
        
        if not row:
            return None
        
        return self._build_model(row)

    def create(self, user: User) -> User:
        """Create a new user"""
        query = """
            INSERT INTO users (email, password_hash, created_at, updated_at)
            VALUES (%s, %s, %s, %s)
            RETURNING id
        """
        cursor = self.db.cursor()
        cursor.execute(
            query,
            (user.email, user.password_hash, user.created_at, user.updated_at)
        )
        user.id = cursor.fetchone()[0]
        self.db.commit()
        cursor.close()
        
        return user

    def update(self, user: User) -> User:
        """Update user"""
        query = """
            UPDATE users
            SET email = %s, password_hash = %s, updated_at = NOW()
            WHERE id = %s
        """
        cursor = self.db.cursor()
        cursor.execute(query, (user.email, user.password_hash, user.id))
        self.db.commit()
        cursor.close()
        
        return user

File: services/ruby/repositories/user_repository.rb
module Repositories
  class UserRepository
    def initialize(db_connection)
      @db = db_connection
    end

    def find(id)
      result = @db.exec_params('SELECT * FROM users WHERE id = $1', [id])
      return nil if result.ntuples.zero?
      
      build_user(result[0])
    end

    def find_by_email(email)
      result = @db.exec_params('SELECT * FROM users WHERE email = $1', [email])
      return nil if result.ntuples.zero?
      
      build_user(result[0])
    end

    def all(limit: 100, offset: 0)
      result = @db.exec_params(
        'SELECT * FROM users ORDER BY created_at DESC LIMIT $1 OFFSET $2',
        [limit, offset]
      )
      
      result.map { |row| build_user(row) }
    end

    def create(user)
      result = @db.exec_params(
        'INSERT INTO users (email, password_hash, created_at, updated_at) VALUES ($1, $2, $3, $4) RETURNING id',
        [user.email, user.password_hash, user.created_at, user.updated_at]
      )
      
      user.id = result[0]['id'].to_i
      user
    end

    def update(user)
      @db.exec_params(
        'UPDATE users SET email = $1, password_hash = $2, updated_at = $3 WHERE id = $4',
        [user.email, user.password_hash, Time.now, user.id]
      )
      
      user.updated_at = Time.now
      user
    end

    def delete(id)
      result = @db.exec_params('DELETE FROM users WHERE id = $1', [id])
      result.cmd_tuples > 0
    end

    def count
      result = @db.exec('SELECT COUNT(*) FROM users')
      result[0]['count'].to_i
    end

    private

    def build_user(row)
      Models::User.new(
        id: row['id'].to_i,
        email: row['email'],
        password_hash: row['password_hash'],
        created_at: Time.parse(row['created_at']),
        updated_at: Time.parse(row['updated_at'])
      )
    end
  end
end
